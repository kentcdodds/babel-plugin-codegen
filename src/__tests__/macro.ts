import path from 'path'
import pluginTester from 'babel-plugin-tester'
import stripAnsi from 'strip-ansi'
import plugin from 'babel-plugin-macros'

const projectRoot = path.join(__dirname, '../../')

expect.addSnapshotSerializer({
  print(val) {
    return stripAnsi(val as string)
      .split(projectRoot)
      .join('<PROJECT_ROOT>/')
  },
  test(val) {
    return typeof val === 'string'
  },
})

// the fixtures is handy because we can also test the type defs at the same time
pluginTester({
  plugin,
  pluginName: 'codegen/macro',
  snapshot: true,
  babelOptions: {filename: __filename, presets: ['@babel/preset-typescript']},
  fixtures: path.join(__dirname, './fixtures/macro'),
})

pluginTester({
  plugin,
  pluginName: 'codegen/macro',
  snapshot: true,
  babelOptions: {filename: __filename, parserOpts: {plugins: ['jsx']}},
  tests: {
    // couldn't figure out how to do type defs for JSX ðŸ˜¬
    // (got an error about not being able to create an overload... I guess it can't be all the things?)
    // PRs welcome
    'as jsx': `
      const Codegen = require('./helpers/codegen.macro')
      const ui = (
        <Codegen>{"module.exports = '<div>Hi</div>'"}</Codegen>
      )
    `,
    'as jsx with tag': `
      const Codegen = require('./helpers/codegen.macro')
      const ui = (
        <Codegen>{\`module.exports = '<div>Hello</div>'\`}</Codegen>
      )
    `,
    'with multiple': `
      import codegen from './helpers/codegen.macro'

      codegen\`module.exports = ['a', 'b', 'c'].map(l => 'export const ' + l + ' = ' + JSON.stringify(l)).join(';')\`
    `,
    'with a single export declaration': `
      import codegen from './helpers/codegen.macro'

      codegen\`module.exports = "export const a = 'a'"\`
    `,
    'with a single import declaration': `
      import codegen from './helpers/codegen.macro'

      codegen\`module.exports = "import a from 'a'"\`
    `,
    'invalid usage: as fn argument': {
      code: `
        import codegen from './helpers/codegen.macro';
        var x = doSomething(codegen);
      `,
      error: true,
    },
    'invalid usage: missing code string': {
      code: `
        import codegen from './helpers/codegen.macro';
        var x = codegen;
      `,
      error: true,
    },
  },
})
