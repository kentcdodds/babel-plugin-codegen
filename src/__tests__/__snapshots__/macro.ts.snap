// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen/macro as jsx with tag: as jsx with tag 1`] = `

const Codegen = require('./helpers/codegen.macro')
const ui = (
  <Codegen>{\`module.exports = '<div>Hello</div>'\`}</Codegen>
)

      ↓ ↓ ↓ ↓ ↓ ↓

const ui = <div>Hello</div>


`;

exports[`codegen/macro as jsx: as jsx 1`] = `

const Codegen = require('./helpers/codegen.macro')
const ui = (
  <Codegen>{"module.exports = '<div>Hi</div>'"}</Codegen>
)

      ↓ ↓ ↓ ↓ ↓ ↓

const ui = <div>Hi</div>


`;

exports[`codegen/macro invalid usage: as fn argument: invalid usage: as fn argument 1`] = `

import codegen from './helpers/codegen.macro';
var x = doSomething(codegen);

      ↓ ↓ ↓ ↓ ↓ ↓

Error: <PROJECT_ROOT>/src/__tests__/macro.ts: ./helpers/codegen.macro: codegen macro must be used as a tagged template literal, function, jsx, or .require call
  1 | import codegen from './helpers/codegen.macro';
> 2 | var x = doSomething(codegen);
    |                     ^^^^^^^

`;

exports[`codegen/macro invalid usage: missing code string: invalid usage: missing code string 1`] = `

import codegen from './helpers/codegen.macro';
var x = codegen;

      ↓ ↓ ↓ ↓ ↓ ↓

Error: <PROJECT_ROOT>/src/__tests__/macro.ts: ./helpers/codegen.macro: codegen macro must be used as a tagged template literal, function, jsx, or .require call
  1 | import codegen from './helpers/codegen.macro';
> 2 | var x = codegen;
    |         ^^^^^^^

`;

exports[`codegen/macro with a single export declaration: with a single export declaration 1`] = `

import codegen from './helpers/codegen.macro'

codegen\`module.exports = "export const a = 'a'"\`

      ↓ ↓ ↓ ↓ ↓ ↓

export const a = 'a'


`;

exports[`codegen/macro with a single import declaration: with a single import declaration 1`] = `

import codegen from './helpers/codegen.macro'

codegen\`module.exports = "import a from 'a'"\`

      ↓ ↓ ↓ ↓ ↓ ↓

import a from 'a'


`;

exports[`codegen/macro with multiple: with multiple 1`] = `

import codegen from './helpers/codegen.macro'

codegen\`module.exports = ['a', 'b', 'c'].map(l => 'export const ' + l + ' = ' + JSON.stringify(l)).join(';')\`

      ↓ ↓ ↓ ↓ ↓ ↓

export const a = 'a'
export const b = 'b'
export const c = 'c'


`;
